File: cppe/src/IceE/Object.cpp
// **********************************************************************
//
// Copyright (c) 2003-2007 ZeroC, Inc. All rights reserved.
//
// This copy of Ice-E is licensed to you under the terms described in the
// ICEE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <IceE/Object.h>
#include <IceE/Incoming.h>
#include <IceE/LocalException.h>
#include <IceE/SafeStdio.h>

using namespace std;
using namespace Ice;
using namespace IceInternal;

IceUtil::Shared* IceInternal::upCast(Object* p) { return p; }

bool
Ice::Object::operator==(const Object& r) const
{
    return this == &r;
}

bool
Ice::Object::operator<(const Object& r) const
{
    return this < &r;
}

Int
Ice::Object::ice_hash() const
{
    return static_cast<Int>(reinterpret_cast<Long>(this) >> 4);
}

static const string __Ice__Object_ids[] =
{
    "::Ice::Object"
};

bool
Ice::Object::ice_isA(const string& s, const Current&) const
{
    return s == __Ice__Object_ids[0];
}

void
Ice::Object::ice_ping(const Current&) const
{
    // Nothing to do.
}

vector<string>
Ice::Object::ice_ids(const Current&) const
{
    return vector<string>(&__Ice__Object_ids[0], &__Ice__Object_ids[1]);
}

const string&
Ice::Object::ice_id(const Current&) const
{
    return __Ice__Object_ids[0];
}

const ::std::string&
Ice::Object::ice_staticId()
{
    return __Ice__Object_ids[0];
}

ObjectPtr
Ice::Object::ice_clone() const
{
    throw CloneNotImplementedException(__FILE__, __LINE__);
}

DispatchStatus
Ice::Object::___ice_isA(Incoming& __inS, const Current& __current)
{
    BasicStream* __is = __inS.is();
    BasicStream* __os = __inS.os();
    string __id;
    __is->read(__id, false);
    bool __ret = ice_isA(__id, __current);
    __os->write(__ret);
    return DispatchOK;
}

DispatchStatus
Ice::Object::___ice_ping(Incoming&, const Current& __current)
{
    ice_ping(__current);
    return DispatchOK;
}

DispatchStatus
Ice::Object::___ice_ids(Incoming& __inS, const Current& __current)
{
    BasicStream* __os = __inS.os();
    vector<string> __ret = ice_ids(__current);
    __os->write(&__ret[0], &__ret[0] + __ret.size(), false);
    return DispatchOK;
}

DispatchStatus
Ice::Object::___ice_id(Incoming& __inS, const Current& __current)
{
    BasicStream* __os = __inS.os();
    string __ret = ice_id(__current);
    __os->write(__ret, false);
    return DispatchOK;
}

string Ice::Object::__all[] =
{
    "ice_id",
    "ice_ids",
    "ice_isA",
    "ice_ping"
};

DispatchStatus
Ice::Object::__dispatch(Incoming& in, const Current& current)
{
    pair<string*, string*> r =
	equal_range(__all, __all + sizeof(__all) / sizeof(string), current.operation);

    if(r.first == r.second)
    {
	throw OperationNotExistException(__FILE__, __LINE__, current.id, current.facet, current.operation);
    }					     

    switch(r.first - __all)
    {
        case 0:
        {
	    return ___ice_id(in, current);
        }
        case 1:
        {
	    return ___ice_ids(in, current);
        }
        case 2:
	{
	    return ___ice_isA(in, current);
	}
	case 3:
	{
	    return ___ice_ping(in, current);
	}
    }

    assert(false);
    throw OperationNotExistException(__FILE__, __LINE__, current.id, current.facet, current.operation);
}

static const char*
operationModeToString(OperationMode mode)
{
    switch(mode)
    {
    case Normal:
	return "::Ice::Normal";

    case Nonmutating:
	return "::Ice::Nonmutating";

    case Idempotent:
	return "::Ice::Idempotent";
    }

    return "???";
}

void
Ice::Object::__invalidMode(OperationMode expected, OperationMode received)
{
    assert(expected != received);

    if(expected == Idempotent && received == Nonmutating)
    {
	// 
	// Fine: typically an old client still using the deprecated nonmutating keyword
	//
	
	//
	// Note that expected == Nonmutating and received == Idempotent is not ok:
	// the server may still use the deprecated nonmutating keyword to detect updates
	// and the client should not break this (deprecated) feature.
	//
    }
    else
    {
	Ice::MarshalException ex(__FILE__, __LINE__);
	ex.reason = Ice::printfToString("unexpected operation mode. expected = %s received = %s", 
					operationModeToString(expected), operationModeToString(received));
	throw ex;
    }
}

DispatchStatus
Ice::Blobject::__dispatch(Incoming& in, const Current& current)
{
    vector<Byte> inParams;
    vector<Byte> outParams;
    Int sz = in.is()->getReadEncapsSize();
    in.is()->readBlob(inParams, sz);
    bool ok = ice_invoke(inParams, outParams, current);
    in.os()->writeBlob(outParams);
    if(ok)
    {
	return DispatchOK;
    }
    else
    {
	return DispatchUserException;
    }
}

File: java/src/Ice/src/main/java/com/zeroc/Ice/Object.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

package com.zeroc.Ice;

import java.util.concurrent.CompletionStage;

import com.zeroc.IceInternal.Incoming;

/**
 * The base interface for servants.
 **/
public interface Object
{
    /** @hidden */
    final static String[] _iceIds =
    {
        "::Ice::Object"
    };

    /**
     * Holds the results of a call to <code>ice_invoke</code>.
     **/
    public class Ice_invokeResult
    {
        /**
         * Default initializes the members.
         **/
        public Ice_invokeResult()
        {
        }

        /**
         * One-shot constructor to initialize the members.
         *
         * @param returnValue True for a succesful invocation with any results encoded in <code>outParams</code>.
         * False if a user exception occurred with the exception encoded in <code>outParams</code>.
         * @param outParams The encoded results.
         **/
        public Ice_invokeResult(boolean returnValue, byte[] outParams)
        {
            this.returnValue = returnValue;
            this.outParams = outParams;
        }

        /**
         * If the operation completed successfully, the return value
         * is <code>true</code>. If the operation raises a user exception,
         * the return value is <code>false</code>; in this case, <code>outParams</code>
         * contains the encoded user exception. If the operation raises a run-time exception,
         * it throws it directly.
         **/
        public boolean returnValue;

        /**
         * The encoded out-paramaters and return value for the operation. The return value
         * follows any out-parameters.
         **/
        public byte[] outParams;
    }

    /**
     * Tests whether this object supports a specific Slice interface.
     *
     * @param s The type ID of the Slice interface to test against.
     * @param current The {@link Current} object for the invocation.
     * @return <code>true</code> if this object has the interface
     * specified by <code>s</code> or derives from the interface
     * specified by <code>s</code>.
     **/
    default boolean ice_isA(String s, Current current)
    {
        return java.util.Arrays.binarySearch(ice_ids(current), s) >= 0;
    }

    /**
     * Tests whether this object can be reached.
     *
     * @param current The {@link Current} object for the invocation.
     **/
    default void ice_ping(Current current)
    {
        // Nothing to do.
    }

    /**
     * Returns the Slice type IDs of the interfaces supported by this object.
     *
     * @param current The {@link Current} object for the invocation.
     * @return The Slice type IDs of the interfaces supported by this object, in base-to-derived
     * order. The first element of the returned array is always <code>::Ice::Object</code>.
     **/
    default String[] ice_ids(Current current)
    {
        return _iceIds;
    }

    /**
     * Returns the Slice type ID of the most-derived interface supported by this object.
     *
     * @param current The {@link Current} object for the invocation.
     * @return The Slice type ID of the most-derived interface.
     **/
    default String ice_id(Current current)
    {
        return ice_staticId();
    }

    /**
     * Returns the Slice type ID of the interface supported by this object.
     *
     * @return The return value is always ::Ice::Object.
     **/
    public static String ice_staticId()
    {
        return _iceIds[0];
    }

    /** @hidden */
    final static String[] _iceOps =
    {
        "ice_id",
        "ice_ids",
        "ice_isA",
        "ice_ping"
    };

    /**
     * Dispatches an invocation to a servant. This method is used by dispatch interceptors to forward an invocation
     * to a servant (or to another interceptor).
     *
     * @param request The details of the invocation.
     * @return A completion stage if the dispatched asynchronously, null otherwise.
     * @throws UserException A user exception that propagates out of this method will be marshaled as the result.
     *
     * @see DispatchInterceptor
     **/
    default CompletionStage<OutputStream> ice_dispatch(Request request)
        throws UserException
    {
        Incoming in = (Incoming)request;
        in.startOver();
        return _iceDispatch(in, in.getCurrent());
    }

    /**
     * @hidden
     * @param in -
     * @param current -
     * @return -
     * @throws UserException -
     **/
    default CompletionStage<OutputStream> _iceDispatch(Incoming in, Current current)
        throws UserException
    {
        int pos = java.util.Arrays.binarySearch(_iceOps, current.operation);
        if(pos < 0)
        {
            throw new OperationNotExistException(current.id, current.facet, current.operation);
        }

        switch(pos)
        {
            case 0:
            {
                return _iceD_ice_id(this, in, current);
            }
            case 1:
            {
                return _iceD_ice_ids(this, in, current);
            }
            case 2:
            {
                return _iceD_ice_isA(this, in, current);
            }
            case 3:
            {
                return _iceD_ice_ping(this, in, current);
            }
        }

        assert(false);
        throw new OperationNotExistException(current.id, current.facet, current.operation);
    }

    /**
     * @hidden
     * @param ostr -
     **/
    default void _iceWrite(OutputStream ostr)
    {
         ostr.startValue(null);
         _iceWriteImpl(ostr);
         ostr.endValue();
    }

    /**
     * @hidden
     * @param ostr -
     **/
    default void _iceWriteImpl(OutputStream ostr)
    {
    }

    /**
     * @hidden
     * @param istr -
     **/
    default void _iceRead(InputStream istr)
    {
         istr.startValue();
         _iceReadImpl(istr);
         istr.endValue(false);
    }

    /**
     * @hidden
     * @param istr -
     **/
    default void _iceReadImpl(InputStream istr)
    {
    }

    /**
     * @hidden
     * @param obj -
     * @param inS -
     * @param current -
     * @return -
     **/
    static CompletionStage<OutputStream> _iceD_ice_isA(Object obj, Incoming inS, Current current)
    {
        InputStream istr = inS.startReadParams();
        String id = istr.readString();
        inS.endReadParams();
        boolean ret = obj.ice_isA(id, current);
        OutputStream ostr = inS.startWriteParams();
        ostr.writeBool(ret);
        inS.endWriteParams(ostr);
        return inS.setResult(ostr);
    }

    /**
     * @hidden
     * @param obj -
     * @param inS -
     * @param current -
     * @return -
     **/
    static CompletionStage<OutputStream> _iceD_ice_ping(Object obj, Incoming inS, Current current)
    {
        inS.readEmptyParams();
        obj.ice_ping(current);
        return inS.setResult(inS.writeEmptyParams());
    }

    /**
     * @hidden
     * @param obj -
     * @param inS -
     * @param current -
     * @return -
     **/
    static CompletionStage<OutputStream> _iceD_ice_ids(Object obj, Incoming inS, Current current)
    {
        inS.readEmptyParams();
        String[] ret = obj.ice_ids(current);
        OutputStream ostr = inS.startWriteParams();
        ostr.writeStringSeq(ret);
        inS.endWriteParams(ostr);
        return inS.setResult(ostr);
    }

    /**
     * @hidden
     * @param obj -
     * @param inS -
     * @param current -
     * @return -
     **/
    static CompletionStage<OutputStream> _iceD_ice_id(Object obj, Incoming inS, Current current)
    {
        inS.readEmptyParams();
        String ret = obj.ice_id(current);
        OutputStream ostr = inS.startWriteParams();
        ostr.writeString(ret);
        inS.endWriteParams(ostr);
        return inS.setResult(ostr);
    }

    /**
     * @hidden
     * @param mode -
     * @return -
     **/
    static String _iceOperationModeToString(OperationMode mode)
    {
        if(mode == OperationMode.Normal)
        {
            return "::Ice::Normal";
        }
        if(mode == OperationMode.Nonmutating)
        {
            return "::Ice::Nonmutating";
        }

        if(mode == OperationMode.Idempotent)
        {
            return "::Ice::Idempotent";
        }

        return "???";
    }

    /**
     * @hidden
     * @param expected -
     * @param received -
     **/
    static void _iceCheckMode(OperationMode expected, OperationMode received)
    {
        if(expected == null)
        {
            expected = OperationMode.Normal;
        }

        if(expected != received)
        {
            if(expected == OperationMode.Idempotent && received == OperationMode.Nonmutating)
            {
                //
                // Fine: typically an old client still using the
                // deprecated nonmutating keyword
                //
            }
            else
            {
                MarshalException ex = new MarshalException();
                ex.reason = "unexpected operation mode. expected = "
                    + _iceOperationModeToString(expected) + " received = "
                    + _iceOperationModeToString(received);
                throw ex;
            }
        }
    }
}
